### PR TITLE
add button for re-plan

### PR DESCRIPTION
**Pull request by CuijieLu**

### CODE REVIEW

File: frontend/src/Plan.js [hunk 1]
```diff
@@ -39,6 +39,10 @@ function Plan() {
         exportExcel(samplesForPooling, columns);
     };
 
+    const handleRePlan = () =>{
+        setGroups(rePlan(samplesForPooling));
```
Comment (DavidStreid) [inline frontend/src/Plan.js:43]: nice

File: frontend/src/util/planRunsUtil.js [hunk 2]
```diff
@@ -0,0 +1,255 @@
+/**
+ * group samples by run type and return a map with runLength as key and sample list as value
+ * 
+ * @param {Array} sampleObjectList array of Sample/PooledSample Objects
+ * @returns a map with runLength as key and sample list as value
+ */
+export function groupSampleByRunType(sampleObjectList){
+    const runTypeList = new Map();
+    for (let i = 0; i < sampleObjectList.length; i++){
```
Comment (DavidStreid) [inline frontend/src/util/planRunsUtil.js:9]: This is fine, but ES6's for-of is also nice - up to you

```
for (const sample of sampleObjectList) {
        if(runTypeList.has(sample.runLength)){
            runTypeList.get(sample.runLength).push(sample);
        }
        ...
}
```

File: frontend/src/util/planRunsUtil.js [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+ * barcode collison checking between two barcodes by given length and number of mismatch allowed
+ * 
+ * @param {string} seq1 barcode
+ * @param {string} seq2 barcode
+ * @param {number} length length that want to perform the collison checking
+ * @param {number} numOfMismatch number of mismatch allowed
+ * @returns true if collision happens under situation of certain number of mismatch allowed
+ */
+function barcodeCollision(seq1, seq2, length, numOfMismatch){
+    var seq1_frag = seq1.substr(0,length);
```
Comment (DavidStreid) [inline frontend/src/util/planRunsUtil.js:29]: I think, `var` -> `const` in general for variables that aren't re-assigned
- "...var declarations are globally scoped or function/locally scoped..."
- "...a variable declared with const remains the same within its scope. It cannot be updated or re-declared..."
https://www.freecodecamp.org/news/var-let-and-const-whats-the-difference/

File: frontend/src/util/planRunsUtil.js [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+            }
+        }
+    } else{
+        for (let i = 0; i < length; i++){
+            if (seq1_frag[i] != seq2_frag[i]){
+                mismatch1 = mismatch1 + 1;
+            }
+        }
+    }               
+    if (mismatch1 <= numOfMismatch * 2 && mismatch2 <= numOfMismatch * 2){
```
Comment (DavidStreid) [inline frontend/src/util/planRunsUtil.js:51]: Can we make a note of where this comes from? Maybe this quote from the bcl2fastq error output when it has a barcode collision - "Barcodes with too few mismatches are ambiguous ( less than 2 times the number of mismatches plus 1)"?




```
By default, bcl2fastq allows 1 mismatch in each barcode. Barcodes with too few mismatches are ambiguous ( less than 2 times the number of mismatches plus 1). To reduce the number of allowed mismatches, use the command line option: '--barcode-mismatches'. Note that particularly for barcodes with only 1 mismatch, there is the danger that some reads will be written to the wrong sample due to errors in the barcode sequence.
```

File: frontend/src/util/planRunsUtil.js [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+ * barcode collison checking between two list of barcodes by given length and number of mismatch allowed
+ * 
+ * @param {Array or string} seqList1 list of barcode or individual barcode
+ * @param {Array or string} seqList2 list of barcode or individual barcode
+ * @param {number} numOfMismatch number of mismatch allowed
+ * @returns true if collision happens under situation of certain number of mismatch allowed
+ */
+function listBarcodeCollision(seqList1, seqList2, numOfMismatch){
+    //merge single barcode/list barcode into one array
+    var emptyList = [];
```
Comment (DavidStreid) [inline frontend/src/util/planRunsUtil.js:69]: `const`

File: frontend/src/util/planRunsUtil.js [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    emptyList.push(seqList1, seqList2);
```
Comment (DavidStreid) [inline frontend/src/util/planRunsUtil.js:70]: Maybe `push`/spread-operator - 
```
l1 = [ 1, 2, 3 ]
l2 = [4, 5, 6]
empty = []
empty.push( ...l1, ...l2 )
console.log(empty)
[1, 2, 3, 4, 5, 6]
```

File: frontend/src/util/planRunsUtil.js [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+    var newSeqList = emptyList.flat();
+    //get minimum barcode length
+    var minLength = 30;
+    for (let i = 0; i < newSeqList.length; i++){
```
Comment (DavidStreid) [inline frontend/src/util/planRunsUtil.js:74]: Good time for `for`/`of`

File: frontend/src/util/planRunsUtil.js [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+        if(newSeqList[i].length < minLength){
+            minLength = newSeqList[i].length;
+        }
+    }
+    //check barcode collision one by one
+    for(let i = 0; i < newSeqList.length - 1; i++){
```
Comment (DavidStreid) [inline frontend/src/util/planRunsUtil.js:80]: Good time for i, j indexes

File: frontend/src/util/planRunsUtil.js [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+            minLength = pool1.barcodeSeq[i].length ;
+        }
+    }
+    for (let i = 0; i < pool2.barcodeSeq.length; i++){
+        if(pool2.barcodeSeq[i].length < minLength){
+            minLength = pool2.barcodeSeq[i].length ;
+        }
+    }
+    for (let i = 0; i < pool1.barcodeSeq.length; i++){
+        for (let j = 0; j < pool2.barcodeSeq.length; j++){
```
Comment (DavidStreid) [inline frontend/src/util/planRunsUtil.js:125]: why not `for(let j = i + 1; j < newSeqList.length; j++){` like you did before? Then you could also get rid of ` && !(i == 0 && j == 0)`, right?

File: frontend/src/util/planRunsUtil.js [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+    var collisionListPool = [];
+    var collisionListLib = [];
+     //function to sort collisionList by collision times
+    function makeSortedCollisionList(collisionList1){
+        var collisionMap = new Map();
+        for(let i = 0; i < collisionList1.length; i++){
+            let count = 0;
+            for(let j = 0; j < collisionList1.length; j++){ 
+                if((i != j) && (sampleBarcodeCollision(collisionList1[i], collisionList1[j],0))){
+                    count = count + 1;
```
Comment (DavidStreid) [inline frontend/src/util/planRunsUtil.js:169]: `count += 1`

File: frontend/src/util/planRunsUtil.js [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+        var collisionListSorted1 = [];
+        for (const key of collisionMapSorted.keys()){
+            collisionListSorted1.push(key);
+        } 
+        return collisionListSorted1;
+    }
+
+    function sortSamplesByProject(collisionListLibrary) {
+        var projectMap = new Map();
+        for (let i = 0; i < collisionListLibrary.length; i++){
```
Comment (DavidStreid) [inline frontend/src/util/planRunsUtil.js:184]: `for`/`of` looping

File: frontend/src/util/rePlan.js [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        if (element.isIncluded === true){
+            element.groupID = "";
+            sampleForRePlan.push(element);
+        }else{
+            element.groupID = "";
+            sampleLeft.push(element);
+        }
+        
+    });
+    var runTypeResult = new Map();
```
Comment (DavidStreid) [inline frontend/src/util/rePlan.js:19]: Can we not use `Map` (maybe just in the future)? 
It seems like maps are used when the keys are objects - https://dmitripavlutin.com/maps-vs-plain-objects-javascript/

**Review by DavidStreid — CHANGES_REQUESTED** [review]
Looks good! Just some syntax things to think about if you want to add them.

One thing that might be a good change is in `checkPoolNormal` (line 125) and the looping

File: frontend/src/util/planRunsUtil.js (hunk 9 above)
Comment (CuijieLu) [reply frontend/src/util/planRunsUtil.js:125]: This is cross check between two lists, so each one need to be compared with the whole list of the other list

File: frontend/src/util/rePlan.js (hunk 12 above)
Comment (CuijieLu) [reply frontend/src/util/rePlan.js:19]: I will look into it, maybe can change with next step

**Review by DavidStreid — APPROVED** [review]
looks good, thanks!

**Pull request merged**